### PR TITLE
Develop the Command Handler

### PR DESCRIPTION
**Pull request by aritosteles**
|Related issue|
|---|
|[#83](https://github.com/wazuh/wazuh-agent/issues/83)|

## Description
This PR aims to create the Command Handler, in charge of removing command messages from the Agent Queue and dispatching them to the corresponding modules for execution.
 

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux

### CODE REVIEW

File: src/agent/configuration_parser/src/configuration_parser.cpp [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace configuration
 
             tbl = toml::parse_str(
                 R"([agent]
-                server_mgmt_api_port = "55000"
+                server_mgmt_api_port = "8080"
```
Comment (TomasTurina) [inline src/agent/configuration_parser/src/configuration_parser.cpp:23]: Don't change this.

File: src/agent/command_store/CMakeLists.txt [hunk 2]
```diff
@@ -9,6 +9,7 @@ include(../../cmake/CommonSettings.cmake)
 set_common_settings()
 
 add_library(CommandStore src/command_store.cpp)
+include(../../cmake/ConfigureTarget.cmake)
```
Comment (TomasTurina) [inline src/agent/command_store/CMakeLists.txt:12]:
```suggestion

include(../../cmake/ConfigureTarget.cmake)
```
Please, add a space before this line.

File: src/agent/include/agent.hpp [hunk 3]
```diff
@@ -26,4 +27,6 @@ class Agent
     AgentInfo m_agentInfo;
     configuration::ConfigurationParser m_configurationParser;
     communicator::Communicator m_communicator;
+    command_handler::CommandHandler m_commandHandler;
+    MultiTypeQueue m_agentQueue;
```
Comment (TomasTurina) [inline src/agent/include/agent.hpp:31]: Duplicated. See `m_messageQueue`.

File: src/agent/src/message_queue_utils.hpp [hunk 4]
```diff
@@ -15,3 +14,7 @@ boost::asio::awaitable<std::string> GetMessagesFromQueue(std::shared_ptr<IMultiT
 void PopMessagesFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, MessageType messageType);
 
 void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const std::string& commands);
+
+std::string GetJsonValue(const nlohmann::json& j, const std::string& key);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:18]: As we talk, there should be a function in the library to do this.

File: src/agent/src/message_queue_utils.hpp [hunk 5]
```diff
@@ -15,3 +14,7 @@ boost::asio::awaitable<std::string> GetMessagesFromQueue(std::shared_ptr<IMultiT
...
+
+bool SetJsonValue(nlohmann::json& j, const std::string& key, const std::string& value);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:20]: Same.

File: src/agent/command_handler/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+set(CMAKE_TOOLCHAIN_FILE "${CMAKE_SOURCE_DIR}/../../vcpkg/scripts/buildsystems/vcpkg.cmake")
+set(VCPKG_MANIFEST_DIR ${CMAKE_SOURCE_DIR}/../../)
+
+project(CommandHandler)
+
+include(../../cmake/CommonSettings.cmake)
+set_common_settings()
+
+add_library(CommandHandler src/command_handler.cpp)
+configure_target(CommandHandler)
```
Comment (TomasTurina) [inline src/agent/command_handler/CMakeLists.txt:12]:
```suggestion

include(../../cmake/ConfigureTarget.cmake)
configure_target(CommandHandler)
```

File: src/agent/src/agent.cpp [hunk 7]
```diff
@@ -43,6 +59,52 @@ void Agent::Run()
         [this]([[maybe_unused]] const std::string& response)
         { PopMessagesFromQueue(m_messageQueue, MessageType::STATELESS); }));
 
+    m_taskManager.EnqueueTask(m_commandHandler.ProcessCommandsFromQueue<command_store::Command>(
+        [this]() -> std::optional<command_store::Command>
```
Comment (TomasTurina) [inline src/agent/src/agent.cpp:63]: Maybe we can call `GetMessagesFromQueue` here instead of this lambda.

Comment (TomasTurina) [inline src/agent/src/agent.cpp:63]: Maybe we can create a function similar to `GetMessagesFromQueue` here instead of this lambda.

File: src/agent/src/agent.cpp [hunk 8]
```diff
@@ -43,6 +59,52 @@ void Agent::Run()
...
+
+            // nlohmann::json jdataData = m.data.at(0).at("data");
+            // for (auto& [key, value] : jdataData.items())
+            // {
+            //     std::cout << key << ": " << value << std::endl;
+            // }
+
+            return cmd;
+        },
+        [this]() -> bool
```
Comment (TomasTurina) [inline src/agent/src/agent.cpp:97]: Same, here we can create and use a similar function than `PopMessagesFromQueue`. Maybe `PopCommandFromQueue`?

Comment (aritosteles) [reply src/agent/src/agent.cpp]: Done

File: src/agent/src/agent.cpp (hunk 7 above)
Comment (aritosteles) [reply src/agent/src/agent.cpp]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 5 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 4 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp]: Done

File: src/agent/configuration_parser/src/configuration_parser.cpp (hunk 1 above)
Comment (aritosteles) [reply src/agent/configuration_parser/src/configuration_parser.cpp]: Done

File: src/agent/command_store/CMakeLists.txt (hunk 2 above)
Comment (aritosteles) [reply src/agent/command_store/CMakeLists.txt:13]: Done

File: src/agent/include/agent.hpp (hunk 3 above)
Comment (aritosteles) [reply src/agent/include/agent.hpp]: Removed

File: src/agent/command_handler/CMakeLists.txt (hunk 6 above)
Comment (aritosteles) [reply src/agent/command_handler/CMakeLists.txt:13]: Done

File: src/agent/command_handler/tests/command_handler_test.cpp
Comment (TomasTurina) [inline src/agent/command_handler/tests/command_handler_test.cpp:1]: Please, add some tests.

File: src/agent/src/agent.cpp [hunk 9]
```diff
@@ -43,6 +44,10 @@ void Agent::Run()
         [this]([[maybe_unused]] const std::string& response)
         { PopMessagesFromQueue(m_messageQueue, MessageType::STATELESS); }));
 
+    m_taskManager.EnqueueTask(m_commandHandler.ProcessCommandsFromQueue<command_store::Command>(
+        [this]() { return GetCommandFromQueue(m_messageQueue); },
+        [this]() { return PopCommandFromQueue(m_messageQueue); },
+        [](command_store::Command& cmd) -> int { return DispatchCommand(cmd); }));
+
     m_signalHandler->WaitForSignal();
-    m_communicator.Stop();
```
Comment (TomasTurina) [inline src/agent/src/agent.cpp:47]: Why is this?

File: src/agent/src/message_queue_utils.cpp [hunk 10]
```diff
@@ -1,15 +1,17 @@
-#include <message_queue_utils.hpp>
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:1]: Remove space.

File: src/agent/sqlite_manager/CMakeLists.txt [hunk 11]
```diff
@@ -12,8 +12,11 @@ find_package(fmt REQUIRED)
 find_package(SQLiteCpp REQUIRED)
 
 add_library(SQLiteManager src/sqlite_manager.cpp)
+
+include(../../cmake/ConfigureTarget.cmake)
 configure_target(SQLiteManager)
-target_include_directories(SQLiteManager PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include PRIVATE ${SQLiteCpp_INCLUDE_DIRS})
+
+target_include_directories(SQLiteManager PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include ${SQLiteCpp_INCLUDE_DIRS})
```
Comment (TomasTurina) [inline src/agent/sqlite_manager/CMakeLists.txt:19]: Is this change necessary? Remove the PRIVATE keyword?

File: src/agent/src/message_queue_utils.hpp [hunk 12]
```diff
@@ -15,3 +17,9 @@ boost::asio::awaitable<std::string> GetMessagesFromQueue(std::shared_ptr<IMultiT
 void PopMessagesFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, MessageType messageType);
 
 void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const std::string& commands);
+
+std::optional<command_store::Command> GetCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue);
+
+bool PopCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue);
+
+int DispatchCommand(command_store::Command&);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:25]: This is not a message queue util. Maybe we can create `command_utils.hpp` and `command_utils.cpp` to place this.

File: src/agent/src/message_queue_utils.hpp [hunk 13]
```diff
@@ -15,3 +17,9 @@ boost::asio::awaitable<std::string> GetMessagesFromQueue(std::shared_ptr<IMultiT
 void PopMessagesFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, MessageType messageType);
 
 void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const std::string& commands);
+
+std::optional<command_store::Command> GetCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:21]: Why don't you return an `awaitable` instead of an `optional`?

File: src/agent/src/message_queue_utils.hpp [hunk 14]
```diff
@@ -15,3 +17,9 @@ boost::asio::awaitable<std::string> GetMessagesFromQueue(std::shared_ptr<IMultiT
...
+
+bool PopCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:23]: Since we don't need to check the return value, I think this function can return void.

File: src/agent/src/message_queue_utils.hpp [hunk 15]
```diff
@@ -1,10 +1,12 @@
 #pragma once
 
+#include <command.hpp>
 #include <message.hpp>
 
 #include <boost/asio/awaitable.hpp>
 
 #include <memory>
+#include <optional>
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:9]: Remove if we change to use `awaitable`.

File: src/agent/src/message_queue_utils.hpp (hunk 12 above)
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.hpp:25]:
```suggestion
int DispatchCommand(const command_store::Command&);
```

File: src/agent/src/message_queue_utils.cpp [hunk 16]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+
+    Message m = multiTypeQueue->getNext(MessageType::COMMAND);
+    nlohmann::json jasonData = m.data.at(0).at("data");
+
+    std::string strId = jasonData["id"].dump();
+
+    // remove quotation marks from id
+    strId.erase(strId.begin());
+    strId.erase(strId.end() - 1);
+    int64_t id = std::stoll(strId);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:69]: I think the ID could be a string. We shouldn't assume it will be only numbers.

File: src/agent/src/message_queue_utils.cpp [hunk 17]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+
+    std::string module = jasonData["origin"]["module"].dump();
+    std::string command = jasonData.contains("command") ? jasonData["command"].dump() : "";
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:72]: This isn't optional, should always be present.

File: src/agent/src/message_queue_utils.cpp [hunk 18]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+
+std::optional<command_store::Command> GetCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue)
+{
+    if (multiTypeQueue->isEmpty(MessageType::COMMAND))
+    {
+        return std::nullopt;
+    }
+
+    Message m = multiTypeQueue->getNext(MessageType::COMMAND);
+    nlohmann::json jasonData = m.data.at(0).at("data");
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:62]:
```suggestion
    nlohmann::json jsonData = m.data.at(0).at("data");
```

File: src/agent/src/message_queue_utils.cpp [hunk 19]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+
+    // remove quotation marks from id
+    strId.erase(strId.begin());
+    strId.erase(strId.end() - 1);
+    int64_t id = std::stoll(strId);
+
+    std::string module = jasonData["origin"]["module"].dump();
+    std::string command = jasonData.contains("command") ? jasonData["command"].dump() : "";
+    std::string parameters = jasonData.contains("parameters") ? jasonData["parameters"].dump() : "";
+    std::string result = "Pending execution";
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:74]: Not necessary.

File: src/agent/src/message_queue_utils.cpp [hunk 20]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+    command_store::Status status = command_store::Status::IN_PROGRESS;
+
+    command_store::Command cmd(id, module, command, parameters, result, status);
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:77]:
```suggestion
    command_store::Command cmd(id, module, command, parameters, "", status);
```

File: src/agent/src/message_queue_utils.cpp [hunk 21]
```diff
@@ -48,3 +50,44 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+bool PopCommandFromQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue)
+{
+    return multiTypeQueue->pop(MessageType::COMMAND);
+}
+
+int DispatchCommand(command_store::Command& cmd)
+{
+    // TO_DO: Implement real dispatch function.
+    std::cout << "Dispatching command " << cmd.m_module << "\n";
+    std::this_thread::sleep_for(std::chrono::milliseconds(DISPATCH_WAIT_MSECS));
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:91]: Remove this line, won't be necessary. Also remove https://github.com/wazuh/wazuh-agent/pull/150/files#diff-6e0f7ef87f491f91b5469ed4abda6d6d12f1470d38e6a0b26c6a7997fd4aa773R7 and https://github.com/wazuh/wazuh-agent/pull/150/files#diff-6e0f7ef87f491f91b5469ed4abda6d6d12f1470d38e6a0b26c6a7997fd4aa773R14.

File: src/agent/command_handler/include/command_handler.hpp [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+            std::unique_ptr<boost::asio::steady_timer> expTimer = std::make_unique<boost::asio::steady_timer>(executor);
+
+            while (true)
+            {
+                auto cmd = GetCommandFromQueue();
+                if (cmd == std::nullopt)
+                {
+                    expTimer->expires_after(1000ms);
+                    co_await expTimer->async_wait(boost::asio::use_awaitable);
+                    std::cout << "Queue is empty - WAITING" << "\n";
```
Comment (TomasTurina) [inline src/agent/command_handler/include/command_handler.hpp:34]: Remove this noisy log.

File: src/agent/command_handler/include/command_handler.hpp [hunk 23]
```diff
@@ -0,0 +1,53 @@
...
+#include <queue>
+
+namespace command_handler
+{
+    class CommandHandler
+    {
+    public:
+        CommandHandler();
+
+        template<typename T>
```
Comment (TomasTurina) [inline src/agent/command_handler/include/command_handler.hpp:17]: Why isn't this function definition in the `.cpp`? Is the `.cpp` even necessary as it is almost empty?

File: src/agent/command_handler/include/command_handler.hpp [hunk 24]
```diff
@@ -0,0 +1,53 @@
...
+                    std::cout << "Queue is empty - WAITING" << "\n";
+                    continue;
+                }
+
+                m_commandStore.StoreCommand(cmd.value());
+                PopCommandFromQueue();
+                DispatchCommand(cmd.value());
+
+                cmd.value().m_status = command_store::Status::SUCCESS;
+                cmd.value().m_result = "Successfully executed";
```
Comment (TomasTurina) [inline src/agent/command_handler/include/command_handler.hpp:43]: These values should be obtained from the result of `DispatchCommand` execution.

File: src/agent/command_handler/include/command_handler.hpp (hunk 23 above)
Comment (aritosteles) [reply src/agent/command_handler/include/command_handler.hpp:17]: The function needs to be in the hpp since it's a template function. The cpp is not really needed, so... gone.

File: src/agent/command_handler/include/command_handler.hpp (hunk 22 above)
Comment (aritosteles) [reply src/agent/command_handler/include/command_handler.hpp:34]: Removed

File: src/agent/sqlite_manager/CMakeLists.txt (hunk 11 above)
Comment (aritosteles) [reply src/agent/sqlite_manager/CMakeLists.txt:19]: I believe it was needed at some point when trying to build the components separately. But since we are leaving that for further down the road, I just reverted the change.

File: src/agent/src/agent.cpp (hunk 9 above)
Comment (aritosteles) [reply src/agent/src/agent.cpp:47]: Reverted.

File: src/agent/src/message_queue_utils.cpp (hunk 10 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:1]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 12 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp:25]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 14 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp:23]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 12 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp:25]: Done

File: src/agent/src/message_queue_utils.cpp (hunk 18 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:62]: Done

File: src/agent/src/message_queue_utils.cpp (hunk 19 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:74]: Removed

File: src/agent/src/message_queue_utils.cpp (hunk 20 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:77]: Done

File: src/agent/src/message_queue_utils.cpp (hunk 21 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:91]: Done

File: src/agent/command_handler/include/command_handler.hpp (hunk 24 above)
Comment (aritosteles) [reply src/agent/command_handler/include/command_handler.hpp]: Done

File: src/agent/src/message_queue_utils.cpp (hunk 17 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp]: Done

File: src/agent/src/message_queue_utils.cpp (hunk 16 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp]: Done

File: src/agent/src/message_queue_utils.cpp [hunk 25]
```diff
@@ -48,3 +46,34 @@ void PushCommandsToQueue(std::shared_ptr<IMultiTypeQueue> multiTypeQueue, const
...
+    {
+        return std::nullopt;
+    }
+
+    Message m = multiTypeQueue->getNext(MessageType::COMMAND);
+    nlohmann::json jsonData = m.data.at(0).at("data");
+
+    std::string id = jsonData["id"].dump();
+
+    // remove quotation marks from id
```
Comment (TomasTurina) [inline src/agent/src/message_queue_utils.cpp:62]: Is this necessary?

Comment (aritosteles) [reply src/agent/src/message_queue_utils.cpp:62]: I found that using ```nlohmann::json::get<std::string>()``` instead of ```dump()``` returns the value without quotation marks. Yet for the parameters member we still need to use ```dump()``` probably because it's an array.

File: src/agent/command_handler/tests/CMakeLists.txt [hunk 26]
```diff
@@ -0,0 +1,5 @@
+find_package(GTest CONFIG REQUIRED)
+
+add_executable(command_handler_test command_handler_test.cpp)
```
Comment (TomasTurina) [inline src/agent/command_handler/tests/CMakeLists.txt:3]:
```suggestion
add_executable(command_handler_test command_handler_test.cpp)
configure_target(command_handler_test)
```

Comment (aritosteles) [reply src/agent/command_handler/tests/CMakeLists.txt:3]: Done

File: src/agent/src/message_queue_utils.hpp (hunk 13 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp:21]: The function is non-blocking since it calls co_await on its internal timer. Let's keep an eye on this and see if it's necessary, we can change it down the road.

File: src/agent/src/message_queue_utils.hpp (hunk 15 above)
Comment (aritosteles) [reply src/agent/src/message_queue_utils.hpp:9]: Answered above.

File: src/agent/command_handler/tests/command_handler_test.cpp
Comment (TomasTurina) [reply src/agent/command_handler/tests/command_handler_test.cpp:1]: Will do as part of https://github.com/wazuh/wazuh-agent/issues/146.
